### PR TITLE
Update the parse function to accept an entity id

### LINKED ISSUE #187: Using the standard parse function for creating entities does not generate an entity_id
the parse_ftm function should include the create of an entity_id value. This should be able to be defined within the yml but if it is not then it should be created from a hash of the  url

### PR DESCRIPTION
**Pull request by Rosencrantz**
This includes a tweak to the parse function so that it generates an entity id before creating an entity. There are two ways in which this can occur

1. Supply a list of xpath values that will be concatenated together and hashed in order to generate a unique key
2. Do nothing and allow the parse function to automatically generate a key based on the url of the page that is being parsed.

### CODE REVIEW

File: docs/buildingcrawler.md [hunk 1]
```diff
@@ -371,8 +371,13 @@ parse:
       author: .//meta[@name="author"]/@content
       publishedAt: .//*[@class="date"]/text()
       description: .//meta[@property="og:description"]/@content
+    keys:
```
Comment (sunu) [inline docs/buildingcrawler.md:374]: I would suggest we use syntax similar to ftm mappings for consistency. Something like https://github.com/alephdata/aleph/blob/main/mappings/md_companies.yml#L15-L17

So the keys section will look like:
```
keys:
  - title
  - author
```

File: memorious/operations/parse.py [hunk 2]
```diff
@@ -91,15 +92,52 @@ def parse_for_metadata(context, data, html):
...
+        for item in keys:
+            temp_key = html.xpath(item["key"])
+            if len(temp_key) > 0:
+                key_string += "".join(temp_key)
+
+        if (
+            hashlib.md5(key_string.encode("utf-8")).hexdigest()
+            != hashlib.md5("".encode("utf-8")).hexdigest()
+        ):
+            entity_id = hashlib.md5(key_string.encode("utf-8")).hexdigest()
```
Comment (sunu) [inline memorious/operations/parse.py]: We can use `make_id` from `memorious.helpers.key` instead of making the key using hashlib.

File: memorious/operations/parse.py [hunk 3]
```diff
@@ -91,15 +92,52 @@ def parse_for_metadata(context, data, html):
...
+
+
 def parse_ftm(context, data, html):
     properties = context.params.get("properties")
     properties_dict = {}
+
     for key, value in properties.items():
         properties_dict[key] = html.xpath(value)
 
+    data["entity_id"] = hashlib.md5(data["url"].encode("utf-8")).hexdigest()
```
Comment (sunu) [inline memorious/operations/parse.py]: It's probably better and more consistent to make keys out of the supplied keys only instead of falling back on a default. We should instead raise an error in case of null keys so that the user can notice and change their keying strategy.

File: memorious/operations/aleph.py [hunk 4]
```diff
@@ -8,69 +9,75 @@
...
-    meta["countries"] = data.get("countries", countries)
-    mime_type = context.params.get("mime_type")
-    meta["mime_type"] = data.get("mime_type", mime_type)
-    return meta
+def _create_document_metadata(meta_base: Meta, context: Context, data: dict) -> Meta:
+    languages: list[str] = list(context.params.get("languages", []))
+    countries: list[str] = list(context.params.get("countries", []))
+    mime_type: str = context.params.get("mime_type", "")
+
+    context.log.warn(languages)
```
Comment (sunu) [inline memorious/operations/aleph.py]: Stray warning

File: memorious/operations/aleph.py [hunk 5]
```diff
@@ -8,69 +9,75 @@
...
+        file_name=data.get("file_name"),
+        retrieved_at=data.get("retrieved_at"),
+        modified_at=data.get("modified_at"),
+        published_at=data.get("published_at"),
+        headers=data.get("headers", {}),
+        keywords=data.get("keywords", []),
+    )
 
     if data.get("aleph_folder_id"):
         meta["parent"] = {"id": data.get("aleph_folder_id")}
```
Comment (sunu) [inline memorious/operations/aleph.py]: We are using 2 different styles here. We are treating meta as a object in the line above and as a dict here. I think we should be consistent and use a single style throughout the function.

And imo we can merge `_create_meta_object` and `_create_document_metadata` together into just one function and only set whatever metadata is available.

File: memorious/logic/meta.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# from typing_extensions import TypedDict
```
Comment (sunu) [inline memorious/logic/meta.py]: IMO this is not generic enough to be its own module. We should put it in the aleph operation module.

File: memorious/operations/parse.py [hunk 7]
```diff
@@ -88,22 +92,41 @@ def parse_for_metadata(context, data, html):
                 if value is not None:
                     data[key] = value
                 break
+    meta_paths.update(data)
```
Comment (sunu) [inline memorious/operations/parse.py]: Why is this necessary?

File: memorious/operations/parse.py [hunk 8]
```diff
@@ -88,22 +92,41 @@ def parse_for_metadata(context, data, html):
...
-def parse_ftm(context, data, html):
+def get_entity_id_from_keys(keys: Optional[list], properties: dict, html) -> str:
+    temp_key = ""
+
+    if isinstance(keys, list):
+        for key in keys:
+            temp_key = "".join(properties[key])
+
+        if not temp_key == "":
+            return make_id(temp_key)
```
Comment (sunu) [inline memorious/operations/parse.py]: `make_id` can take multiple arguments. You don't need to join them beforehand.

File: docs/buildingcrawler.md (hunk 1 above)
Comment (sunu) [reply docs/buildingcrawler.md:374]: The `keys` section needs to be updated now I think?

File: example/config/simple_article_scraper.yml [hunk 9]
```diff
@@ -45,6 +45,10 @@ pipeline:
         author: .//meta[@name="author"]/@content
         publishedAt: .//*[@class="date"]/text()
         description: .//meta[@property="og:description"]/@content
+      keys:
```
Comment (sunu) [inline example/config/simple_article_scraper.yml:48]: May be the method needs to be changed to use the built-in `parse` method instead of a custom Python method to match the documentation example?

File: memorious/logic/meta.py (hunk 6 above)
Comment (Rosencrantz) [reply memorious/logic/meta.py]: But it's not a module. It's the definition of a type. I would type definitions seperate from modules that actually do stuff.

File: example/config/simple_article_scraper.yml (hunk 9 above)
Comment (Rosencrantz) [reply example/config/simple_article_scraper.yml:48]: Sure. The problem with doing that though is that the scraper won't be able to extract the body of the article, which is why the custom script exists. I guess, as it's an example it doesn't really matter too much, but that is why we have a difference.

Personally I don't have an issue with having the documentation not match the example in the repo

File: memorious/logic/meta.py (hunk 6 above)
Comment (sunu) [reply memorious/logic/meta.py]: Putting the type definition near the relevant code is the convention we're already using in other places. See for example https://github.com/alephdata/followthemoney/blob/master/followthemoney/schema.py#L24

Comment (Rosencrantz) [reply memorious/logic/meta.py]: Fair enough. I'll move the meta file into the operations module.

Comment (sunu) [reply memorious/logic/meta.py]: I actually meant the Meta type definition should live in operations/aleph.py and not in a separate file since it's very specific to the aleph_emit operations and unlikely to be reused anywhere else.

File: memorious/operations/meta.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# from typing_extensions import TypedDict
+from typing import Optional, TypedDict
+
+
+class MetaBase(TypedDict):
```
Comment (sunu) [inline memorious/operations/meta.py]: MetaBase vs Meta difference is no longer used. So I guess it's safe to merge these two together now?

**Comment by sunu** [comment]
Could you fix the merge conflict too? I think it's from the changes Simon made to fix a couple of issues in `aleph_emit` operation.

File: memorious/operations/aleph.py [hunk 11]
```diff
@@ -9,59 +11,84 @@
...
-    mime_type = context.params.get("mime_type")
-    meta["mime_type"] = data.get("mime_type", mime_type)
-    return meta
-
-
-def _create_meta_object(context, data) -> dict:
+from memorious.logic.context import Context
+
+
+class Meta(MetaBase, total=False):
```
Comment (sunu) [inline memorious/operations/aleph.py]: MetaBase is undefined

File: memorious/operations/aleph.py [hunk 12]
```diff
@@ -9,59 +11,84 @@
...
+    content_hash: Optional[str] = data.get("content_hash")
+    source_url: str = data.get("source_url", data.get("url"))
+    foreign_id: str = data.get("foreign_id", data.get("request_id", source_url))
+    document_id: Optional[str] = context.get_tag(
+        make_key(collection_id, foreign_id, content_hash)
+    )
+
+    if document_id:
         context.log.info("Skip aleph upload: %s", foreign_id)
         data["aleph_id"] = document["id"]
```
Comment (sunu) [inline memorious/operations/aleph.py:88]: `document` is undefined

**Review by sunu — CHANGES_REQUESTED** [review]
Looks like the merge conflict wasn't fixed the way it should be. So the code is left in a broken state. A test for the aleph operations should have caught this. But we don't have any tests for it. May be it's worth adding one now since we are making changes to the aleph_emit operations.

**Comment by Rosencrantz** [comment]
This died
